msyslog-v1.08a+smac/src/man/om_mysql.8
.\"	$CoreSDI: om_mysql.8,v 1.1.2.2.4.13 2001/11/20 09:56:23 alejo Exp $
.\"
.\" Copyright (c) 2001
.\"	Core-SDI SA. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Core-SDI SA nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Jun 13, 2000
.Dt OM_MYSQL 8
.Os Core-SDI
.Sh NAME
.Nm MySQL output module
.Nd
.Xr syslogd 8
output module used to log on MySQL servers
.Sh SYNOPSIS
.Nm mysql
.Op Fl D
.Op Fl s Ar host<:port>
.Op Fl u Ar username
.Op Fl p Ar password
.Op Fl d Ar database
.Op Fl t Ar table
.Op Fl F
.Op Fl P
.Sh DESCRIPTION
.Nm MySQL output module 
receives a message as an ascii string and logs it on a MySQL server. For
proper initialization, the following parameters should be set:
.Bl -tag -width Ds
.It Fl s Ar hostname<:port>
Specify the MySQL server hostname and optionally the port.
.It Fl u Ar username
Specifies the username to use on the MySQL server. This user
must have enough permissions to insert on the specified database and table.
.It Fl p Ar password
The plaintext password for this MySQL user. Thus it is recommended NOT to
leave
.Xr syslog.conf 5
world readable if this module is used.
.It Fl d Ar database
The database name to use on the specified MySQL server.
.It Fl t Ar table
The table name to use on the specified MySQL database.
.It Fl D
Do DELAYED inserts. See your MySQL documentation.
.It Fl F
Insert the facility level name on the table on a field named "facility"
.It Fl P
Insert the priority level name on the table on a field named "priority"
.Sh EXAMPLES
The table should have a format like this:
.Pp
.Bd -literal
  CREATE TABLE syslogTB (
    facility char(10),    # OPTIONAL field for facility
    priority char(10),    # OPTIONAL field for priority
    date date,           # date of this log message
    time time,           # time of this message
    host varchar(128),   # host logging
    message  text,        # message
    seq  int unsigned auto_increment primary key # optional sequence number
  ); 
.Ed
.Pp
The field names should be respected. The type of fields
.Va host
and
.Va msg
can be any text format smaller than 128 and 1024 chars respectively.
.Pp
The field seq may be needed for some
.Xr audit 1
features.
.Sh SEE ALSO
.Xr syslog 3 ,
.Xr syslog.conf 5 ,
.Xr om_classic 8 ,
.Xr om_peo 8 ,
.Xr om_pgsql 8 ,
.Xr om_regex 8 ,
.Xr om_tcp 8 ,
.Xr om_udp 8 ,
.Xr peochk 8 ,
.Xr syslogd 8
.Sh BUGS
.Bl -bullet
.It
Since the MySQL module is used to connect with username and password,
care must be pointed on file permissions and that user permissions
on the MySQL server.
.It
Submit bugs at this project's Sourceforge Bug reporting system at:
http://sourceforge.net/tracker/?func=add&group_id=25741&atid=385117
You may also report them directly to the authors; send an email to
core.devel.alat@corest.com, describing the problem the most you can,
containing also machine description, hardware description, the
configuration file (/etc/syslog.conf), the OS description, and the
invoking command line.
The more you describe the bug, the faster we can fix it.
.El


html/ChangeLog.txt
$Id$

Feature Request:

Bugs:
1.  Change code to allow for tailing spaces when looking at the expression field

3/10/2004
	-  Removing any reference to company name

3/9/2004
	-  Changed out company logo
	-  Added GPL notice to all libraries

1/21/2004
	-  Found a bug where a host would email off an alert with no text.  I've put a stop gap fix in.

1/13/2004
        -  Fixed bug with msyslog.pgsql where SET STORAGE didn't have a space before it.
        -  Fixed bug where hour in time of rules would default to 18:00 for no good reason
	-  Fixed two problems with cloning rules:  1)  order was not preserved  2)  an imported version of the database couldn't clone because fields that were empty needed to have non-null defaults applied
	-  Forgot to merge in latest VPN reports

1/6/2004
        -  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/6/2004
	-  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/1/2004
	-  Fixed spelling error in with the word 'threshold' in the rules schema
	-  Fixed a bug in vacuumdb where it was 'ANALYZ' not 'ANALYZE'

12/19/2003
	-  Updated code to look into /opt/apache

12/10/2003
	-  Updated database to schema to not use compressed text fields(we'll see how this performs)
	-  working on adding interfaces to more lock data and other new stats with PostgreSQL V7.4
	-  cleaned up some button descriptions on the maintenance page, also added a lock view as well as a settings view

12/9/2003
	-  Finished rule.php support for basic timer maintenance.  Need to add another page to graft timers onto rules.
	-  processlogs.php is now setup to support rule timers
	-  need to write rule expiration process 
	-  started converting away from compressed text in hopes of providing faster data retrieval.... also pulled OIDs from DB definition

12/8/2003
	-  Adding support for date and time based rules with date ranges, day of week selections, as well as deleted rules, need to add interface to control rule timestamp properties

12/7/2003
	-  finished adding accumulation thresholds

12/3/2003
	-  adding support for both types of thresholds.  Need to update web pages to reflect new radio buttons.
	-  added web configuration support for supression thresholds and accumulating thresholds, now onto updating the log processor
	-  basic supression works(kinda), need to verify functionality

12/2/2003
	-  customer profiles can now have multiple hosts added at once.
	-  fixed a stupid bug where <prev|refresh|next> didn't behave right.  next would stop working if you started at oct-26-2003 and it would stay on oct-26-2003
	-  Added web-based framework and database schema to support alert supression thresholds

12/1/2003
	-  Updated processlogs.php code to better deal with single entry/no rules vs mutliple rules

11/19/2003
	-  Updated mail table to enforce unique login ids(effectively stopping two processes from running at the same time, one will crash and die(safely))
	-  Updated openmail and closeopenmail to use transaction support since PostgreSQL no longer does server-side auto-commit(ie. convert everything to transactions)
	-  Transaction support should now be officially added, will do some testing

11/18/2003
	-  Updating program for support with PostGreSQL V7.4
	-  Fixed host process table to reflect the fact that multiple hosts are in there by default	

10/13/2003
	-  Added alert total to bottom of alert aggregation

10/6/2003
	-  finished support for alert aggregation
	-  updated 1stview to pull the current time and date
	-  fixed host.php bug where you could expire syslogs but not be forced to expire syslogs

10/5/2003
	-  still adding support for alert aggregation.  Basic aggregation works plus alert zooming but need to add support for across the board for other alert queries

10/3/2003
	-  added ability to un/suspend log processors from the web interface
	-  updated maintenance to rebuild all indexes in an better manner(ie. grab the index list from the DB rather then by hand)
	-  added additional framework to do alert aggregation interface.  Need to add 'aggregation code' for display

9/24/2003
	-  weeklyindexrebuild.php now pulls all indexes from the system and rebuilds them.  The result is that the system will now rebuild any new indexes without manual reconfiguration.

9/15/2003
	-  launchid was not initiated correctly in the clonerule.

9/2/2003
	-  Fixed a bug where '\'s at the end of a line caused problems because we were not properly dealing with them in general.  Fixed that.

8/23/2003	
	-  Created another bug when fixing 5000 line paging.  Timestamp was thrown off in view.php

8/18/2003	
	-  Syncing changes from production smt environment:  vacuumtsyslog.php
	-  Updated weeklyindexrebuild.php to account for the correct indexes
	-  Updated maint.php to account for the three new indexes for the launch program section

8/13/2003
	-  view.php has had several updates.  Paging should now be fixed.  Multiple searches appeared not to be working correctly.
	-  Needed to add lastid as hidden var if the variable was set
	-  Needed to use urlencode on top of htmlspecialcharacters, filters were broken because of it

8/11/2003
	-  1stfilter.php doesn't list 'global' filters that you down own
	-  modified view.php to not let the user save a filter with no description
	-  added support to delete all of a user's filters(ie. do before delete)
	-  another problem popped up with filter.php when I added the delete user filter option

8/1/2003	
	-  processlogs.php now supports launching external programs!

7/31/2003
	-  Updated vacuumdb.php to do a full vacuum of the TSyslog table.  Why?  Because the system doesn't reclaim disk space or use old delete space for some reason
	-  Almost finished adding launch program code, need to test.

7/27/2003
	-  Added weeklyindexrebuild.php which rebuilds all indexes at 5am Sunday morning

7/23/2003

msyslog-v1.08a+smac/src/modules/om_mysql.c
/*	$CoreSDI: om_mysql.c,v 1.36.2.4.2.3.4.33 2001/11/30 23:13:45 alejo Exp $	*/

/*
 * Copyright (c) 2001, Core SDI S.A., Argentina
 * All rights reserved
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither name of the Core SDI S.A. nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * om_mysql -- MySQL database support Module
 *
 * Author: Alejo Sanchez for Core-SDI SA
 *
 */

#include "config.h"

#include <sys/types.h>
#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif
#include <sys/param.h>
#include <signal.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define SYSLOG_NAMES
#include <syslog.h>
#include <unistd.h>
#include <dlfcn.h>
#include "../modules.h"
#include "../syslogd.h"
#include "sql_misc.h"

/* size of  query buffer */
#define MAX_QUERY	8192
/* how many seconds to wait to give again the error message */
#define MSYSLOG_MYSQL_ERROR_DELAY	30

struct om_mysql_ctx {
	void	*h;
	int	lost;
	char	*table;
	char	*host;
	int	port;
	char	*user;
	char	*passwd;
	char	*db;
	void	*lib;
	int	flags;
	int	(*mysql_ping)(void *);
	void *	(*mysql_init)(void *);
	void *	(*mysql_real_connect)(void *, char *, char *, char *,
	    char *, int, void *, int);
	int	(*mysql_query)(void *, char *);
	void	(*mysql_close)(void *);
	char *	(*mysql_error)(void *);
};
#define	OM_MYSQL_DELAYED_INSERTS	0x2
#define	OM_MYSQL_FACILITY		0x4
#define	OM_MYSQL_PRIORITY		0x8

int om_mysql_close(struct filed *, void *);
char	*decode_val(int, CODE *);

/*
 * Define our prototypes for MySQL functions
 */

#define MYSQL_PORT 3306


int
om_mysql_write(struct filed *f, int flags, struct m_msg *m, void *ctx)
{
	struct om_mysql_ctx *c;
	char	query[MAX_QUERY], err_buf[100], facility[16], priority[16];
	int i;
	RETSIGTYPE (*sigsave)(int);

	dprintf(MSYSLOG_INFORMATIVE, "om_mysql_write: entering [%s] [%s]\n",
	    m->msg, f->f_prevline);

	c = (struct om_mysql_ctx *) ctx;

	/* ignore sigpipes   for mysql_ping */
	sigsave = place_signal(SIGPIPE, SIG_IGN);

	if ( ((c->mysql_ping)(c->h)) != 0 && (((c->mysql_init)(c->h) == NULL)
	    || ((c->mysql_real_connect)(c->h, c->host, c->user, c->passwd, c->db,
	    c->port, NULL, 0)) == NULL) ) {

		/* restore previous SIGPIPE handler */
		place_signal(SIGPIPE, sigsave);
		c->lost++;
		if (c->lost == 1) {
			snprintf(err_buf, sizeof(err_buf), "om_mysql_write: "
			    "Lost connection! [%s]", c->mysql_error?
			    c->mysql_error(c->h) : "<unknown>");
			dprintf(MSYSLOG_SERIOUS, "%s", err_buf);
			logerror(err_buf);
		}
		return (1);
	}

	/* restore previous SIGPIPE handler */
	place_signal(SIGPIPE, sigsave);

	/*
	 * NOTE: could use prioritynames[] and facilitynames[]
	 */
	if (c->flags & OM_MYSQL_FACILITY)
		snprintf(facility, sizeof(facility), "'%s',", decode_val(m->fac<<3, facilitynames));
	if (c->flags & OM_MYSQL_PRIORITY)
		snprintf(priority, sizeof(priority), "'%s',", decode_val(m->pri, prioritynames));

	/* table, yyyy-mm-dd, hh:mm:ss, host, msg  */ 
	i = snprintf(query, sizeof(query), "INSERT %sINTO %s (%s%s date, time, "
	    "host, message) VALUES(%s%s '%.4d-%.2d-%.2d', '%.2d:%.2d:%.2d', '%s', '",
	    (c->flags & OM_MYSQL_DELAYED_INSERTS)? "DELAYED " : "", c->table,
	    (c->flags & OM_MYSQL_FACILITY)? "facility, " : "",
	    (c->flags & OM_MYSQL_PRIORITY)? "priority, " : "",
	    (c->flags & OM_MYSQL_FACILITY)? facility : "",
	    (c->flags & OM_MYSQL_PRIORITY)? priority : "",
	    f->f_tm.tm_year + 1900, f->f_tm.tm_mon + 1, f->f_tm.tm_mday,
	    f->f_tm.tm_hour, f->f_tm.tm_min, f->f_tm.tm_sec, f->f_prevhost);

	if (c->lost) {
		int pos = i;

		/*
		 * Report lost messages, but 2 of them are lost of
		 * connection and this one (wich we are going
		 * to log anyway)
		 */
		snprintf(err_buf, sizeof(err_buf), "om_mysql_write: %i "
		    "messages were lost due to lack of connection",
		    c->lost - 2);

		/* count reset */
		c->lost = 0;

		/* put message escaping special SQL characters */
		pos += to_sql(query + pos, err_buf, sizeof(query) - pos);

		/* finish it with "')" */
		query[pos++] =  '\'';
		query[pos++] =  ')';
		if (pos < sizeof(query))
			query[pos]   =  '\0';
		else
			query[sizeof(query) - 1] = '\0';

		dprintf(MSYSLOG_INFORMATIVE2, "om_mysql_write: query [%s]\n",
		    query);

		if ((c->mysql_query)(c->h, query) < 0)
			return (-1);
	}

	/* put message escaping special SQL characters */
	i += to_sql(query + i, m->msg, sizeof(query) - i);

	/* finish it with "')" */
	query[i++] =  '\'';
	query[i++] =  ')';
	if (i < sizeof(query))
		query[i]   =  '\0';
	else
		query[sizeof(query) - 1] = '\0';

	dprintf(MSYSLOG_INFORMATIVE2, "om_mysql_write: query [%s]\n",
	    query);

	if ((i = (c->mysql_query)(c->h, query)) < 0) {
		snprintf(err_buf, sizeof(err_buf), "om_mysql_write: error "
		    "inserting on table [%s]", c->mysql_error?
		    c->mysql_error(c->h) : "<unknown>");
		dprintf(MSYSLOG_SERIOUS, "%s\n", err_buf);
		return (-1);
	}

	return (1);
}

/*
 *  INIT -- Initialize om_mysql
 *
 *  Parse options and connect to database
 *
 *  params:
 *         -s <host:port>
 *         -u <user>
 *         -p <password>
 *         -b <database_name>
 *         -t <table_name>
 *         -P
 *         -F
 *
 */

int
om_mysql_init(int argc, char **argv, struct filed *f, char *prog, void **c,
    char **status)
{
	struct om_mysql_ctx	*ctx;
	char	*p, err_buf[256], statbuf[1024];
	int	ch;

	if (argv == NULL || *argv == NULL || argc < 2 || f == NULL ||
	    c == NULL || *c != NULL)
		return (-1);

	dprintf(MSYSLOG_INFORMATIVE, "om_mysql_init: alloc context\n");
	/* alloc context */
	if ((*c = (void *) calloc(1, sizeof(struct om_mysql_ctx))) == NULL)
		return (-1);
	ctx = (struct om_mysql_ctx *) *c;

	if ((ctx->lib = dlopen("libmysqlclient.so", DLOPEN_FLAGS)) == NULL) {
		dprintf(MSYSLOG_SERIOUS, "om_mysql_init: Error loading"
		    " api library, %s\n", dlerror());
		free(ctx);  
		return (-1);
	}

	if ( !(ctx->mysql_ping = (int(*)(void *)) dlsym(ctx->lib,
	    SYMBOL_PREFIX "mysql_ping")) || !(ctx->mysql_init =
	    (void * (*)(void*)) dlsym(ctx->lib, SYMBOL_PREFIX "mysql_init"))
	    || !(ctx->mysql_real_connect = (void *(*)(void *, char *, char *,
	    char *, char *, int, void *, int)) dlsym(ctx->lib, SYMBOL_PREFIX
	    "mysql_real_connect"))
	    || !(ctx->mysql_query = (int (*)(void *, char *)) dlsym(ctx->lib,
	    SYMBOL_PREFIX "mysql_query"))
	    || !(ctx->mysql_close = (void (*)(void *)) dlsym(ctx->lib,
	    SYMBOL_PREFIX "mysql_close")) ) {
		dprintf(MSYSLOG_SERIOUS, "om_mysql_init: Error resolving"
		    " api symbols, %s\n", dlerror());
		free(ctx);  
		return (-1);
	}

	/* this may be missing on old versions */
	ctx->mysql_error = (char * (*)(void *)) dlsym(ctx->lib, SYMBOL_PREFIX
	    "mysql_error");

	/* parse line */
	optind = 1;
#ifdef HAVE_OPTRESET
	optreset = 1;
#endif
	while ((ch = getopt(argc, argv, "s:u:p:d:t:DPF")) != -1) {
		switch (ch) {
		case 's':
			/* get database host name and port */
			if ((p = strstr(optarg, ":")) == NULL) {
				ctx->port = MYSQL_PORT;
			} else {
				*p = '\0';
				ctx->port = atoi(++p);
			}
			ctx->host = strdup(optarg);
			break;
		case 'u':
			ctx->user = strdup(optarg);
			break;
		case 'p':
			ctx->passwd = strdup(optarg);
			break;
		case 'd':
			ctx->db = strdup(optarg);
			break;
		case 't':
			ctx->table = strdup(optarg);
			break;
		case 'D':
			ctx->flags |= OM_MYSQL_DELAYED_INSERTS;
			break;
		case 'P':
			ctx->flags |= OM_MYSQL_FACILITY;
			break;
		case 'F':
			ctx->flags |= OM_MYSQL_PRIORITY;
			break;
		default:
			goto om_mysql_init_bad;
		}
	}

	if (ctx->user == NULL || ctx->db == NULL || ctx->port == 0 ||
	    ctx->host == NULL || ctx->table == NULL)
		goto om_mysql_init_bad;

	/* connect to the database */
	if (! (ctx->h = (ctx->mysql_init)(NULL)) ) {

		snprintf(err_buf, sizeof(err_buf), "om_mysql_init: Error "
		    "initializing handle");
		logerror(err_buf);
		goto om_mysql_init_bad;
	}

	dprintf(MSYSLOG_INFORMATIVE, "om_mysql_init: mysql_init returned %p\n",
	    ctx->h);

	dprintf(MSYSLOG_INFORMATIVE, "om_mysql_init: params %p %s %s %s %i"
	    " \n", ctx->h, ctx->host, ctx->user, ctx->db, ctx->port);

	snprintf(statbuf, sizeof(statbuf), "om_mysql: sending "
	    "messages to %s, database %s, table %s.", ctx->host,
	    ctx->db, ctx->table);
	*status = strdup(statbuf);

	if (!((ctx->mysql_real_connect)(ctx->h, ctx->host, ctx->user,
	    ctx->passwd, ctx->db, ctx->port, NULL, 0)) ) {

		snprintf(err_buf, sizeof(err_buf), "om_mysql_init: Error "
		    "connecting to db server [%s], [%s:%i] user [%s] db [%s]",
		    ctx->mysql_error? ctx->mysql_error(ctx->h) : "<unknown>",
		    ctx->host, ctx->port, ctx->user, ctx->db);
		logerror(err_buf);
		return (1);
	}

	return (1);

om_mysql_init_bad:
	if (ctx) {
		om_mysql_close(f, ctx);
		free(ctx);
		*c = NULL;
	}

	*status = NULL;

	return(-1);
}


int
om_mysql_close(struct filed *f, void *ctx)
{
	struct om_mysql_ctx *c;

	c = (struct om_mysql_ctx*) ctx;

	(c->mysql_close)(c->h);

	if (c->table)
		free(c->table);
	if (c->host)
		free(c->host);
	if (c->user)
		free(c->user);
	if (c->passwd)
		free(c->passwd);
	if (c->db)
		free(c->db);
	if (c->lib)
		dlclose(c->lib);

	return (0);
}
